client/state/data-layer/wpcom/read/tags/index.js

```diff
@@ -0,0 +1,73 @@
+/**
+ * External dependencies
+ */
+import { map } from 'lodash';
+
+/**
+ * Internal dependencies
+ */
+import {
+	READER_TAGS_REQUEST,
+} from 'state/action-types';
+import {
+	receiveTags,
+} from 'state/reader/tags/items/actions';
+
+import { http } from 'state/data-layer/wpcom-http/actions';
+import { dispatchRequest } from 'state/data-layer/wpcom-http/utils';
+
+export function requestTags( store, action, next ) {
+	const path = action.payload && action.payload.slug
+		? `/read/tags/${ action.payload.slug }`
+		: '/read/tags';
+
+	store.dispatch( http( {
+		path,
+		method: 'GET',
+		apiVersion: '1.2',
+		onSuccess: action,
+		onFailure: action,
+	} ) );
+
+	next( action );
+}
+
+/**
+ * Normalize response from the api so whether we get back a single tag or a list of tags
+ * we always pass forward a list
+ * @param  {Tag|Tags} apiResponse api response from the tags endpoint
+ * @return {Tags}             An object containing list of tags
+ */
+function fromApi( apiResponse ) {
+	let tags;
+	if ( apiResponse.tag )	 {
+		tags = [ apiResponse.tag ];
+	} else if ( apiResponse.tags ) {
+		tags = map( apiResponse.tags, tag => ( { ...tag, is_following: true } ) );
+	} else {
+		if ( process.env.NODE_ENV === 'development' ) {
+			throw new Error( 'bad api response for /read/tags' );
+		}
+		tags = [];
+	}
+
+	return { tags };
```

Comment:
this is a very imperative way of assigning `tags` and I can't say that I can follow it very easily

it appears like the API might return _either_ a single tag _or_ a list of tags having a different structure. is this right?

also a big point of the `fromApi` function is showing how the API format of data maps into Calypso data, by simply passing along the data here we hide that and don't benefit from it

```js
const validate = ( { tag, tags } ) => {
	if ( ! tag && ! Array.isArray( tags ) ) {
		throw new TypeError(
			`Expected to receive tag information from API but none was found.\n` +
			`Expected either 'tag' as a tag object (but found '${ typeof tag }' instead)\n` +
			`or 'tags' as a list of tag objects (but found '${ typeof tags }' instead)`
		);
	}
}

const fromApi = ( { tag, tags } ) => compact(
	[ tag, ...tags ].map( tag => ( {
		name: tag.name,
		otherTagProperty: tag.some_thing,
		isFollowing: true,
	} ) )
);
```